### PR TITLE
Fix SEPA for QR

### PR DESCRIPTION
**Pull request by cristinamghita**

### CODE REVIEW

File: backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java [hunk 1]
```diff
@@ -661,18 +665,26 @@ else if (!Check.isEmpty(line.getSwiftCode(), true))
...
 				final GenericAccountIdentification1CH othr = objectFactory.createGenericAccountIdentification1CH();
 				id.setOthr(othr);
-
-				final String otherAccountIdentification = line.getOtherAccountIdentification();
-				if (!Check.isEmpty(otherAccountIdentification, true))
+				if (Check.isEmpty(otherAccountIdentification, true) && Check.isEmpty(accountNo, true))
 				{
-					// task 07789
-					othr.setId(otherAccountIdentification); // for task 07789, this needs to contain the ESR TeilehmerNr or PostkontoNr
+					othr.setId(iban.replaceAll(" ", ""));
```
Comment (metas-ts) [inline backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java:675]: i think iban might be `null` at this place

File: backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java [hunk 2]
```diff
@@ -948,7 +960,7 @@ private String extractPaymentType(@NonNull final I_SEPA_Export_Line line)
 				de.metas.payment.esr.model.I_C_BP_BankAccount.class);
 
 		final String paymentMode;
-		if (bPBankAccount.isEsrAccount() && !Check.isEmpty(line.getStructuredRemittanceInfo(), true))
+		if (bPBankAccount.isEsrAccount() && !Check.isEmpty(line.getStructuredRemittanceInfo(), true) && bPBankAccount.getQR_IBAN() == null)
```
Comment (metas-ts) [inline backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java:963]:
```suggestion
		if (bPBankAccount.isEsrAccount() && !Check.isEmpty(line.getStructuredRemittanceInfo(), true) && Check.isBlank(bPBankAccount.getQR_IBAN()))
```

File: backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java [hunk 3]
```diff
@@ -1025,7 +1037,7 @@ private String getBPartnerNameById(final int bpartnerRepoId)
 	@VisibleForTesting
 	static boolean isInvalidQRReference(final String reference)
 	{
-		if(reference.length() != 27)
+		if(reference!=null && reference.length() != 27)
```
Comment (metas-ts) [inline backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java:1040]: i think if reference is null we get an NPE further down

**Review by metas-ts — CHANGES_REQUESTED** [review]

File: backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java [hunk 4]
```diff
@@ -661,18 +663,21 @@ else if (!Check.isEmpty(line.getSwiftCode(), true))
...
-				final String otherAccountIdentification = line.getOtherAccountIdentification();
-				if (!Check.isEmpty(otherAccountIdentification, true))
-				{
-					// task 07789
-					othr.setId(otherAccountIdentification); // for task 07789, this needs to contain the ESR TeilehmerNr or PostkontoNr
-				}
-				else
 				{
-					othr.setId(bankAccount.getAccountNo());
+					if (!Check.isEmpty(otherAccountIdentification, true))
```
Comment (metas-ts) [inline backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java:672]:
```suggestion
					if (Check.isNotBlank(otherAccountIdentification))
```

File: backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java [hunk 5]
```diff
@@ -661,18 +663,21 @@ else if (!Check.isEmpty(line.getSwiftCode(), true))
...
 				id.setOthr(othr);
-
-				final String otherAccountIdentification = line.getOtherAccountIdentification();
-				if (!Check.isEmpty(otherAccountIdentification, true))
-				{
-					// task 07789
-					othr.setId(otherAccountIdentification); // for task 07789, this needs to contain the ESR TeilehmerNr or PostkontoNr
-				}
-				else
 				{
```
Comment (metas-ts) [inline backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java:671]: i think we don't need these branckets, or am i missing sth?

File: backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java [hunk 6]
```diff
@@ -948,7 +953,7 @@ private String extractPaymentType(@NonNull final I_SEPA_Export_Line line)
 				de.metas.payment.esr.model.I_C_BP_BankAccount.class);
 
 		final String paymentMode;
-		if (bPBankAccount.isEsrAccount() && !Check.isEmpty(line.getStructuredRemittanceInfo(), true))
+		if (bPBankAccount.isEsrAccount() && !Check.isEmpty(line.getStructuredRemittanceInfo(), true) && Check.isBlank(bPBankAccount.getQR_IBAN()))
```
Comment (metas-ts) [inline backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java:956]:
```suggestion
		if (bPBankAccount.isEsrAccount() && Check.isNotBlank(line.getStructuredRemittanceInfo()) && Check.isBlank(bPBankAccount.getQR_IBAN()))
```

File: backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java [hunk 7]
```diff
@@ -1025,7 +1030,7 @@ private String getBPartnerNameById(final int bpartnerRepoId)
 	@VisibleForTesting
 	static boolean isInvalidQRReference(final String reference)
 	{
-		if(reference.length() != 27)
+		if(reference!=null && reference.length() != 27)
```
Comment (metas-ts) [inline backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java:1033]: wdyt about 
```java
	static boolean isInvalidQRReference(@NonNull final String reference)
	{
		if(reference.length() != 27)
```
?

**Review by metas-ts — CHANGES_REQUESTED** [review]

**Comment by adi-stefan** [comment]
merged to master in https://github.com/metasfresh/metasfresh/commit/1a5447da0089f0037a3e24628e6bec1cfc21b8f7

**Pull request merged**